README.md
# iobeam Python Library

**[iobeam](https://iobeam.com)** is a data platform for connected devices.

This is a Python library for sending data to **iobeam**.
For more information on the iobeam, please [check our our documentation](https://docs.iobeam.com).

*Please note that we are currently invite-only. You will need an invite
to generate a valid token and use our APIs.
([Sign up here](https://iobeam.com) for an invite.)*


## Before you start

Before you can start sending data to the iobeam backend, you'll need a
`project_id` and  `project_token` (with write-access enabled) for a valid
**iobeam** account. You can get these easily with our
[command-line interface (CLI) tool](https://github.com/iobeam/iobeam) or by
accessing your project settings from [our web app](https://app.iobeam.com).

You need python **2.7.9+** or **3.4.3+** (earlier versions of python3 may
work, but it has not been tested).


## Installation

The easiest way to install is to use `pip`:

    pip install iobeam

#### Installing from source

You will need the `requests` and `pyjwt` libraries installed.
You can get it via pip:

    pip install requests pyjwt

If you are using Python2, or Python3 earlier than 3.5,
you will need the `enum34` library as well:

    pip install enum34

Then, to install from source:

    git clone https://github.com/iobeam/iobeam-client-python.git

Then make sure that the `iobeam` folder is in your `PYTHONPATH`.


## Overview

This library allows Python clients to send and query data to the
iobeam backend.

At a high-level, here's how it works:

1. Build an iobeam client object with your `project_id` and
`project_token`

1. Make sure your device is registered, either generating a `device_id` in
code or via another method (e.g., our CLI or REST APIs).

1. Create `iobeam.DataStore` objects for your data streams.
You can create one object per stream, or, if you have streams that are always
collected together (e.g. the axes of a gyroscope), you can put them in
one object.

1. Add data values to your `iobeam.DataStore` objects.

1. When you're ready, send your data to the iobeam backend

1. Optionally, you can use this library to retrieve your data.


## Getting Started

Here's how to get started, using a basic example that sends temperature
data to iobeam. (For simplicity, let's assume that the current temperature
can be accessed with `getTemperature()`).

(Reminder: Before you start, create a user account, project, and
project_token (with write access) using the iobeam APIs, CLI or web app.
Write down your new `project_id` and `project_token`.)

### iobeam Initialization

There are several ways to initialize the iobeam client. All require
that you have `project_id` and `project_token` before hand.

**Without a registered `device_id`**

Perhaps the most natural way is to let the device register itself.
There are two ways to register a `device_id`:

(1) Let iobeam generate one for you:
```python
from iobeam import iobeam

...

builder = iobeam.ClientBuilder(PROJECT_ID, PROJECT_TOKEN) \
                .saveToDisk().registerDevice()
iobeamClient = builder.build()
```

(2) Provide your own (must be unique to your project):
```python
from iobeam import iobeam

...

builder = iobeam.ClientBuilder(PROJECT_ID, PROJECT_TOKEN) \
                .saveToDisk().registerDevice(deviceId="my_desired_id")
iobeamClient = builder.build()
```

With the `saveToDisk()` call, the `device_id` will be saved to disk in the
directory the script is called from (optionally, you can supply a `path`).
On future calls, this on-disk storage will be read first.
If a `device_id` exists, the `registerDevice` will do nothing; otherwise,
it will get a new random ID from us. If you provide a _different_ `device_id` to `registerDevice`, the old one will be replaced.

**With a registered `device_id`**

If you have registered a `device_id` (e.g. using our
[CLI](https://github.com/iobeam/iobeam)), you can pass this to the `Builder` instead
of registering:
```python
from iobeam import iobeam

...

builder = iobeam.ClientBuilder(PROJECT_ID, PROJECT_TOKEN) \
                .saveToDisk().setDeviceId(DEVICE_ID)
iobeamClient = builder.build()
```

You *must* have registered some other way (CLI, website, previous
installation, etc) for this to work.

**Advanced: not saving to disk**

If you don't want the `device_id` to be automatically stored for you, simply
exclude the `saveToDisk()` call while building:
```python
builder = iobeam.ClientBuilder(PROJECT_ID, PROJECT_TOKEN).registerDevice()
iobeamClient = builder.build()
```

This is useful for cases where you want to persist the ID yourself (e.g.
in a settings file), or if you are making `Iobeam` objects that are
temporary. For example, if the device you are using acts as a relay or
proxy for other devices, it could get the `device_id` from those devices
and have no need to save it.

### Tracking Time-series Data

For a more in-depth discussion about adding data, please see [our guide on
adding data](docs/DataGuide.md).

To track time-series data, you need to decide how to break down your data
streams into "stores", a collection of data streams grouped together. You
create a `iobeam.DataStore` with a list of stream names that it contains.
So if you're tracking just temperature in a store called `conditions`:
```python
conditions = iobeamClient.createDataStore(["temperature"])
```

Then for every data point, you'll want to add it to the store:
```python
now = ...  # e.g., now = int(time.time() * 1000) (import time first)
t = getTemperature()
conditions.add(iobeam.Timestamp(now), {"temperature": t})
```

The values are passed in via a dictionary, keyed by which data stream the
individual values belong to along with a timestamp.
_(For more on timestamps, see the next section.)_

The `iobeam.DataStore` object can hold several streams at once. For
example, if you also had a `getHumidity()` function, you could track both in
the same `DataStore`:
```python
conditions = iobeamClient.createDataStore(["temperature", "humidity"])

now = ... # current time
temp = getTemperature()
humidity = getHumidity()
conditions.add(iobeam.Timestamp(now), {"temperature": temp, "humidity": humidity})
```

Not every `add()` call needs all streams to have a value; if a stream is
omitted from the dictionary, it will be assumed to be `None`.

### Timestamps

By default, if you pass just an integer for timestamp when
constructing an `iobeam.DataPoint`, it will set the unit as
milliseconds. To specify other precisions, you need to use the
`iobeam.Timestamp` and `iobeam.TimeUnit` classes:
```python
# Timestamps in seconds:
now = int(time.time())
ts = iobeam.Timestamp(now, unit=iobeam.TimeUnit.SECONDS)
dp = iobeam.DataPoint(t, timestamp=ts)

# Another way to do milliseconds:
now = int(time.time() * 1000)
ts = iobeam.Timestamp(now, unit=iobeam.TimeUnit.MILLISECONDS)
dp = iobeam.DataPoint(t, timestamp=ts)

# Timestamps in microseconds:
now = int(time.time() * 1000000)
ts = iobeam.Timestamp(now, unit=iobeam.TimeUnit.MICROSECONDS)
dp = iobeam.DataPoint(t, timestamp=ts)
```

Currently we support expressing timestamps in seconds, milliseconds,
and microseconds.

### Connecting to the iobeam backend

You can send your data stored in `iobeam.Iobeam` to the iobeam backend
easily:
```python
iobeamClient.send()
```

This call is blocking and will attempt to send all your data. It will
return `True` if successful.


### Full Sending Example

Here's the full source code for our example:
```python
from iobeam import iobeam
import time

# Constants initialization
PROJECT_ID = ... # int
PROJECT_TOKEN = ... # String

...

# Init iobeam
builder = iobeam.ClientBuilder(PROJECT_ID, PROJECT_TOKEN) \
                .saveToDisk().registerDevice()
iobeamClient = builder.build()
conditions = iobeamClient.createDataStore(["temperature", "humidity"])

...

# Data gathering
now = ... # current time
ts = iobeam.Timestamp(now)

temp = getTemperature()
humidity = getHumidity()
conditions.add(ts, {"temperature": temp, "humidity": humidity})

...

# Data transmission
iobeamClient.send()
```

## Retrieving Data

Once you've sent data to the iobeam backend, you may want to query it and
process it. To do that, you'll need to create an `iobeam.QueryReq`, which is
composed of three parts: a project id, a device name (optional), and a series
name (optional). If a series name is not given, all series for that device are
retrieved. If a device name is not given, all devices will be queried.

In the simplest form, here is how you make a few different queries:
```python
# all series from all devices in project PROJECT_ID
q = iobeam.QueryReq(PROJECT_ID)

# all series from device DEVICE_ID in project PROJECT_ID
q = iobeam.QueryReq(PROJECT_ID, deviceId=DEVICE_ID)

# series "temp" from device DEVICE_ID in project PROJECT_ID
q = iobeam.QueryReq(PROJECT_ID, deviceId=DEVICE_ID, seriesName="temp")
```

Then to actually execute the query:
```python
# token is a project token with read access
res = iobeam.makeQuery(token, q)
```

Your result will look something like this:

    { "result": [
        {
            "project_id": "<PROJECT_ID>",
            "device_id": "<DEVICE_ID>",
            "name": "temp",
            "data": [
                {
                    "time":  1427316488000,
                    "value": 22.23
                },
                {
                    "time":  1427316489000,
                    "value": 22.22
                }, ...
            ]
        }, ...
      ],
      "timefmt": "msec"
    }


### Adjusting Your Query

You can modify your with several parameters, such as `to` and `from` and limits
on the values you're interested in:
```python
# start with this basic query
q = iobeam.QueryReq(PROJECT_ID, deviceId=DEVICE_ID, seriesName="temp")

# Last 5 results after a given START:
q = q.limit(5).fromTime(START)

# All results between two times, START and END
q = q.fromTime(START).toTime(END)
# OR...
q = q.inTimeRange(START, END)

# All results where the value is greater than 0
q = q.greaterThan(0)
```

By default, time values are treated as milliseconds. If you'd like to
use a different (seconds or microseconds) you can initialize the
query like this:
```python
q = iobeam.QueryReq(PROJECT_ID, timeUnit=iobeam.TimeUnit.MICROSECONDS)
```

The full list of (chainable) parameters:

    limit(limit)
    fromTime(time)
    toTime(time)
    inTimeRange(startTime, endTime)  # combines previous two
    greaterThan(value)
    lessThan(value)
    inValueRange(min, max)  # combines previous two
    equals(value)


## Running tests

The tests use the Python `unittest` module, along with `mock`. If you are running a Python lower
than 3.3, you need to install `mock` from PyPI:

    pip install mock

(Versions >= 3.3 already have mock installed.)

You can invoke the tests with:

    python -m unittest discover

Or for more verbose output:

    python -m unittest discover -v


CHANGELOG.md
# iobeam Python Library

[iobeam](https://iobeam.com) is a data platform for connected devices.

This is a Python library for sending data to **iobeam**.
For more information on the iobeam, please [check our our documentation](https://docs.iobeam.com).

Please note that we are currently invite-only. You will need an invite
to generate a valid token and use our APIs.
([Sign up here](https://iobeam.com) for an invite.)


## Changelog

### 07/15/16 - v0.11.0
- Client will reuse a DataStore with the same columns when using `createDataStore` if 
it exists (rather than create a duplicate)

### 06/01/16 - v0.10.5
- Clients will reuse HTTP connections more intelligently

### 05/27/16 - v0.10.4
- Fix publishing/build process to include all dependencies

### 03/24/16 - v0.10.3
- Work with both base64 and base64url tokens

### 03/14/16 - v0.10.2
- Check that series names are not empty strings

### 03/07/16 - v0.10.1
- Library checks that device ID is valid with regex `[a-zA-Z0-9:_-]+` instead of error
from server

### 02/01/16 - v0.10.0
- Error messages from the server are now included in stacktrace output
- Errors during refreshing tokens now raise an `UnknownCodeError`

### 01/28/16 - v0.9.7
- Rejects reserved column names, case-insensitive

### 01/27/16 - v0.9.6
- Add additional reserved column name: 'all'
- Fix error with when manually adding `DataStore` to a client (not recommended)

### 01/26/16 - v0.9.5
- Raise `ValueError` if `DataStore` contains reserved column names: time, time_offset

### 01/08/16 - v0.9.4
- Remove accidental dependency on pypandoc when installing from sources

### 12/22/15 - v0.9.2
- `registerOrSetId()` now accepts optional `deviceName` argument

### 12/18/15 - v0.9.1
- **IMPORTANT**: The client now correctly tracks batches it has seen before
so subsequent calls to `send()` do not send nothing.
- Documentation fixes

### 12/16/15 - v0.9.0
- Adding data should now be done via the new `iobeam.createDataStore(columns)` method,
which allows you to track multiple streams of data in one object.
- Internally, old methods involving `DataPoint`s and `DataSerie`s have been converted
to use `DataStore`s.
- Additional documentation provided [here](https://github.com/iobeam/iobeam-client-python/blob/master/docs/DataGuide.md).

### 12/02/15 - v0.8.0
- Automatic refresh of project tokens when they are expired. It is recommended that you
update as soon as possible to avoid breakage.
- bugfix: device ids are no longer allowed to be non-strings when registering

### 10/20/15 - v0.7.0
- **IMPORTANT**: previously sent data points were being kept and sent on subsequent sends, this
has been fixed
- Query assumes time values given in to/from are same unit as when it is initialized

### 10/13/15 - v0.6.0
- `Timestamp`: changed `type` param name to `unit` in constructor
- utils: Adds a check for whether this is python2 or 3
- New Error for duplicate ids
- New `registerOrSetId()` added to `iobeam.ClientBuilder` to allow you to register a name, or
if it already exists, set the client to use that id.

### 10/05/15 - v0.5.2
- Fix missing return for `iobeam.makeQuery()`

### 10/02/15 - v0.5.1

- Use persistent connections for HTTP to improve performance
- Rename `iobeam.MakeQuery()` to `iobeam.makeQuery()`


iobeam/utils/utils.py
"""Common utility functions."""
import jwt
import logging
import re
import sys
import time

IS_PY3 = sys.version_info >= (3, 0)

# For compatibility with both Python 2 and 3.
# pylint: disable=redefined-builtin,invalid-name
if IS_PY3:
    unicode = str
    long = int
# pylint: enable=redefined-builtin,invalid-name

EXPIRY_FUDGE = 1000 * 60 * 60 * 24  # one day in milliseconds

def isExpiredToken(projectToken):
    """Check if token's expiry date (minus a bit) has passed.

    Token is expired if the current time is after the expiry date less EXPIRY_FUDGE.

    Params:
        projectToken - JWT token to decode

    Returns:
        True if expired; False otherwise.

    Raises:
        ValueError - projectToken is not a valid JWT token
    """
    checkValidProjectToken(projectToken)
    opts = {"verify_signature": False, "verify_exp": False}
    try:
        decoded = jwt.decode(projectToken.replace("+", "-").replace("/", "_"),
                             "", options=opts)
    except jwt.DecodeError:
        raise ValueError("invalid jwt token")
    exp = int(decoded["exp"]) * 1000
    now = int(time.time() * 1000)

    return now >= (exp - EXPIRY_FUDGE)


def checkValidProjectId(projectId):
    """Check that a projectId is valid: a position int.

    Raises:
        ValueError - If projectId is None, not an int, or <= 0.
    """
    if projectId is None or not isinstance(projectId, (int, long)):
        raise ValueError("projectId must be an int")
    elif not projectId > 0:
        raise ValueError("projectId must be greater than 0")

def __checkNon0LengthString(value, valueName):
    """Check that a value is a non-0 length string.

    Raises:
        ValueError - If `token` is (a) None, (b) not a string/unicode, or
                     (c) is of length 0.
    """
    if value is None:
        raise ValueError("{} must be a string".format(valueName))
    elif not isinstance(value, (str, unicode)):
        raise ValueError("{} must be a string".format(valueName))
    elif len(value) == 0:
        raise ValueError("{} must be more than 0 characters".format(valueName))

__RESERVED_COL_NAMES = ["time", "time_offset", "all"]

def checkValidSeriesName(name):
    """Check that a series name is valid.

    A valid series name is not reserved by iobeam (time, time_offset, all) and
    a non-0 length string.

    Raises:
        ValueError - If `name` is (a) None, (b) not a string/unicode, (c)
                     is of length 0, or (d) reserved.
    """
    __checkNon0LengthString(name, "columns")
    if name.lower() in __RESERVED_COL_NAMES:
        raise ValueError("'{}' is a reserved column name".format(name))

def checkValidDeviceId(deviceId):
    """Check that a deviceId is valid: string of len > 0

    Raises:
        ValueError - If `deviceId` is (a) None, (b) not a string/unicode, or
                     (c) is of length 0.
    """
    __checkNon0LengthString(deviceId, "deviceId")
    pattern = re.compile("^[a-zA-Z0-9_:-]+$")
    if not pattern.match(deviceId):
        raise ValueError("deviceId can only include a-z, A-Z, 0-9, _, :, and -")


def checkValidProjectToken(token):
    """Check that a token is valid: string of len > 0

    Raises:
        ValueError - If `token` is (a) None, (b) not a string/unicode, or
                     (c) is of length 0.
    """
    __checkNon0LengthString(token, "token")

__LOGGER = None

def getLogger():
    """Get the logger for this library."""

    global __LOGGER
    if __LOGGER is not None:
        logger = __LOGGER
    else:
        logger = logging.getLogger("iobeam")
        logger.setLevel(logging.WARNING)
        __LOGGER = logger

    if len(logger.handlers) == 0:
        logger.addHandler(logging.NullHandler())
    return logger
